### PR TITLE
Adds Modafinil properly this time.

### PR DESCRIPTION
**Pull request by yorii**
:cl: Modafinil
add: Adds new medicine chem that suppresses sleep and very lightly reduces stunrates, has a very low metabolic rate which is randomized and a low overdose treshold. Overdosing is a lethal oxyloss unless treated. (With epipen urgently and with charcoal/calomel before it puts you to sleep)
/:cl:

[why]: Because this adds an in my opinion needed effect to chemistry that the bartender gets almost for free. Chem recipe should be pretty balanced and the effect itself is basically the same as drinking cafe latte plus a quarter of what nicotine gives for stunrates.

Requires: #34903 or #34893 (TRAIT_SLEEPIMMUNE not included in this PR)

(Closed old PR #34610 because I entirely fucked up the branch and don't know how to fix it)

### CODE REVIEW

**Comment by yorii** [comment]
Travis fail due to pre-required PR #34903:
`
code/modules/reagents/chemistry/reagents/medicine_reagents.dm:1318:error: TRAIT_SLEEPIMMUNE: undefined var
`

**Comment by Supermichael777** [comment]
SMOKE

**Comment by DaxDupont** [comment]
This will make great kill juice

**Comment by yorii** [comment]
@DaxDupont not really, a single epipen and you're safe for 80 cycles, basic chloralhydrate is more effective than this. If you want some new kill juice though look no further than #34579 ;)

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 1]
```diff
@@ -1302,3 +1302,72 @@
...
+	description = "Long-lasting sleep suppressant that very slightly reduces stun and knockdown times. Overdosing has horrendous side effects and deals lethal oxygen damage, will knock you unconscious if not dealt with."
+	reagent_state = LIQUID
+	color = "#BEF7D8" // palish blue white
+	metabolization_rate = 0.1 * REAGENTS_METABOLISM
+	overdose_threshold = 10 // low overdose treshold due to very low and random metabolism rate
+	taste_description = "salt" // it actually does taste salty
+	var/overdose_progress = 0 // to track overdose progress
+
+/datum/reagent/medicine/modafinil/on_mob_add(mob/living/M)
+	M.add_trait(TRAIT_SLEEPIMMUNE)
```
Comment (XDTM) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: add_trait(TRAIT_SLEEPIMMUNE, id)

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 2]
```diff
@@ -1302,3 +1302,72 @@
...
+	..()
+
+/datum/reagent/medicine/modafinil/on_mob_delete(mob/living/M)
+	M.remove_trait(TRAIT_SLEEPIMMUNE)
```
Comment (XDTM) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: remove_trait(TRAIT_SLEEPIMMUNE, id)

**Review by XDTM — CHANGES_REQUESTED** [review]
You need unique ids! For chems i usually use the chem's own id, which is unique to the chem and descriptive.

**Review by vuonojenmustaturska — CHANGES_REQUESTED** [review]
Please add trait IDs as suggested by @XDTM

**Comment by yorii** [comment]
Thanks for the reminder on Discord, I was busy celebrating my 30th birthday this weekend and am still a bit drunk I think?

I made the changes and also upped the OD to 15u, because 10u during my last testing was simply way too low with the RNG added to the OD/metabolism. It could probably even be increased to 20u without any balance issues, but I have tested 15u thoroughly and it's pretty okay. (it's not easy whatsoever to be on this all the time, and I assume people will be testing that, enjoy falling over in the hallways)

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 3]
```diff
@@ -1361,7 +1361,7 @@
 				M.Knockdown(20, 1, 0) // you should be in a bad spot at this point unless epipen has been used
 		if(81)
 			to_chat(M, "You're knocked out from the exhaustion!") // at this point you will eventually die unless you get charcoal
-			M.remove_trait(TRAIT_SLEEPIMMUNE)
+			M.remove_trait(TRAIT_SLEEPIMMUNE) //Supposed to remove all sleep immune traits.
```
Comment (XDTM) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: I don't really approve, you might gain the trait from a source that shouldn't be removed.

Comment (yorii) [reply code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: Oh shit you're right @XDTM I should just use the new ignore_sleepimmune on the Sleeping function..

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 4]
```diff
@@ -1361,12 +1361,11 @@
 				M.Knockdown(20, 1, 0) // you should be in a bad spot at this point unless epipen has been used
 		if(81)
 			to_chat(M, "You're knocked out from the exhaustion!") // at this point you will eventually die unless you get charcoal
-			M.remove_trait(TRAIT_SLEEPIMMUNE) //Supposed to remove all sleep immune traits.
-			M.Sleeping(100, 0)
+			M.Sleeping(100, 0, 1)
```
Comment (XDTM) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: TRUE

Comment (duncathan) [reply code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: ^

**Comment by duncathan** [comment]
@yorii

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 5]
```diff
@@ -1302,3 +1302,71 @@
...
+			if(prob(20))
+				to_chat(M, "You have a sudden fit!")
+				M.emote("moan")
+				M.Knockdown(20, 1, 0) // you should be in a bad spot at this point unless epipen has been used
+		if(81)
+			to_chat(M, "You're knocked out from the exhaustion!") // at this point you will eventually die unless you get charcoal
+			M.Sleeping(100, 0, 1)
+			M.adjustOxyLoss(1.5*REM, 0)
+			M.adjustStaminaLoss(1.5*REM, 0)
+		if(82 to INFINITY)i
```
Comment (F-OS) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: bad

**Comment by yorii** [comment]
I've commited the required changes @optimumtact but it's not showing here, probably because closed?
Can you reopen this?

**Review by yorii — COMMENTED** [review]
I did the changes will this give someone a hilight?

**Pull request merged**